// @slack/interactive-messages/dist/util.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorCodes = undefined;
exports.promiseTimeout = promiseTimeout;
exports.packageIdentifier = packageIdentifier;

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

var _package = require('../package.json');

var _package2 = _interopRequireDefault(_package);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function escape(s) {
  return s.replace('/', ':').replace(' ', '_');
}

var errorCodes = exports.errorCodes = {
  PROMISE_TIMEOUT: 'SLACKMESSAGEUTIL_PROMISE_TIMEOUT'
};

function promiseTimeout(ms, promise) {
  // Create a promise that rejects in <ms> milliseconds
  var timeout = new Promise(function (resolve, reject) {
    var id = setTimeout(function () {
      clearTimeout(id);
      var error = new Error('Promise timed out');
      error.code = errorCodes.PROMISE_TIMEOUT;
      reject(error);
    }, ms);
  });
  // Returns a race between our timeout and the passed in promise
  return Promise.race([promise, timeout]);
}

// NOTE: before this can be an external module:
// 1. are all the JS features supported back to a reasonable version?
//    default params, template strings, computed property names
// 2. access to `pkg` will change
// 3. tests
// there will potentially be more named exports in this file
// eslint-disable-next-line import/prefer-default-export
function packageIdentifier() {
  var addons = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var identifierMap = Object.assign({
    [`${_package2.default.name}`]: _package2.default.version,
    [`${_os2.default.platform()}`]: _os2.default.release(),
    node: process.version.replace('v', '')
  }, addons);
  return Object.keys(identifierMap).reduce(function (acc, k) {
    return `${acc} ${escape(k)}/${escape(identifierMap[k])}`;
  }, '');
}
